lxd/seccomp/seccomp.go

```diff
@@ -1675,6 +1692,110 @@ func (s *Server) HandleSchedSetschedulerSyscall(c Instance, siov *Iovec) int {
+func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
+	l := logger.AddContext(logger.Log, logger.Ctx{
+		"container":             c.Name(),
+		"project":               c.Project(),
+		"syscall_number":        siov.req.data.nr,
+		"audit_architecture":    siov.req.data.arch,
+		"seccomp_notify_id":     siov.req.id,
+		"seccomp_notify_flags":  siov.req.flags,
+		"seccomp_notify_pid":    siov.req.pid,
+		"seccomp_notify_fd":     siov.notifyFd,
+		"seccomp_notify_mem_fd": siov.memFd,
+	})
+
+	defer l.Debug("Handling sysinfo syscall")
+
+	runtime.LockOSThread()
+
+	info := unix.Sysinfo_t{}
+	err := unix.Sysinfo(&info)
+	if err != nil {
+		l.Warn("Failed getting sysinfo", logger.Ctx{"err": err})
+		C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+
+		return 0
+	}
+
+	l.Info("tomp Sysinfo", logger.Ctx{"sysinfo": fmt.Sprintf("%+v", info)})
+
+	cg, err := c.CGroup()
+	if err != nil {
+		l.Warn("Failed getting instance cgroup", logger.Ctx{"err": err})
+		C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+
+		return 0
+	}
+
+	l.Info("tomp CGroup")
+
+	// Get Memory limit.
+	memoryLimit, err := cg.GetEffectiveMemoryLimit()
+	if err != nil {
+		l.Warn("Failed getting effective memory limit", logger.Ctx{"err": err})
+
+		C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+		return 0
+	}
+
+	l.Info("tomp GetEffectiveMemoryLimit", logger.Ctx{"memoryLimit": memoryLimit})
+
+	// Get memory usage.
+	memoryUsage, err := cg.GetMemoryUsage()
+	if err != nil {
+		l.Warn("Failed to get memory usage", logger.Ctx{"err": err})
+		C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+
+		return 0
+	}
+
+	l.Info("tomp GetMemoryUsage", logger.Ctx{"memoryUsage": memoryUsage})
+
+	if memoryLimit > 0 {
+		info.Totalram = uint64(memoryLimit)
+		info.Freeram = info.Totalram - uint64(memoryUsage)
+
+		/*
+			out.AddSamples(metrics.MemoryMemTotalBytes, metrics.Sample{Value: float64(memoryLimit)})
+			out.AddSamples(metrics.MemoryMemAvailableBytes, metrics.Sample{Value: float64(memoryLimit - memoryUsage + memoryCached)})
+			out.AddSamples(metrics.MemoryMemFreeBytes, metrics.Sample{Value: float64(memoryLimit - memoryUsage)})
+		*/
+	}
+
+	// Handle swap.
+	/*if s.s.OS.CGInfo.Supports(cgroup.MemorySwapUsage, cg) {
+		swapUsage, err := cg.GetMemorySwapUsage()
+		if err != nil {
+			l.Warn("Failed getting swap usage", logger.Ctx{"err": err})
+			C.seccomp_notify_update_response(siov.resp, 0, C.uint32_t(seccompUserNotifFlagContinue))
+
+			return 0
+		}
+
+		info.Totalswap = uint64(swapUsage)
+	}*/
+
+	// Get writable pointer to buffer of sysinfo syscall result.
+	const sz = int(unsafe.Sizeof(info))
+	var b []byte = (*(*[sz]byte)(unsafe.Pointer(&info)))[:]
+
+	l.Info("tomp Sysinfo write", logger.Ctx{"sysinfo": fmt.Sprintf("%+v", info)})
+
+	_, err = unix.Pwrite(siov.memFd, b, int64(siov.req.data.args[0]))
```

Comment:
I wonder if it would be easy to just defer this pointer work to a C helper here since pointer modification should be simpler and the pwrite() call clearer...